--- a/Makefile.custom	2010-04-23 10:05:48.476350544 -0700
+++ b/Makefile.custom	2010-04-23 10:10:24.194451175 -0700
@@ -107,7 +107,7 @@ bigdata: busybox_unstripped
 
 # Documentation Targets
 .PHONY: doc
-doc: docs/busybox.pod docs/BusyBox.txt docs/BusyBox.1 docs/BusyBox.html
+doc: docs/busybox.pod docs/BusyBox.txt docs/busybox.1 docs/BusyBox.html
 
 # FIXME: Doesn't belong here
        cmd_doc =
@@ -134,10 +134,10 @@ docs/BusyBox.txt: docs/busybox.pod
 	$(Q)-mkdir -p docs
 	$(Q)-pod2text $< > $@
 
-docs/BusyBox.1: docs/busybox.pod
+docs/busybox.1: docs/busybox.pod
 	$(disp_doc)
 	$(Q)-mkdir -p docs
-	$(Q)-pod2man --center=BusyBox --release="version $(KERNELVERSION)" $< > $@
+	$(Q)-pod2man --center=busybox --release="version $(KERNELVERSION)" $< > $@
 
 docs/BusyBox.html: docs/busybox.net/BusyBox.html
 	$(disp_doc)
