--- a/init/init.c	2010-11-30 09:15:30.772252114 -0800
+++ b/init/init.c	2010-11-30 09:19:00.906350490 -0800
@@ -562,6 +562,8 @@ static void run_actions(int action_type)
 	for (a = init_action_list; a; a = a->next) {
 		if (!(a->action_type & action_type))
 			continue;
+		if (a->terminal[0] && access(a->terminal, R_OK | W_OK))
+			continue;
 
 		if (a->action_type & (SYSINIT | WAIT | ONCE | CTRLALTDEL | SHUTDOWN)) {
 			pid_t pid = run(a);
