--- a/scripts/gen_build_files.sh	2010-11-08 15:28:12.000000000 -0800
+++ b/scripts/gen_build_files.sh	2010-11-08 15:50:28.899064030 -0800
@@ -48,7 +48,7 @@ if test x"$new" != x"$old"; then
 fi
 
 # (Re)generate */Kbuild and */Config.in
-{ cd -- "$srctree" && find . -type d; } | while read -r d; do
+{ cd -- "$srctree" && find -name .\?\* -prune -or -type d -print; } | while read -r d; do
 	d="${d#./}"
 
 	src="$srctree/$d/Kbuild.src"
