--- a/scripts/gen_build_files.sh	2010-11-30 09:19:12.450465769 -0800
+++ b/scripts/gen_build_files.sh	2010-11-30 09:19:24.650587598 -0800
@@ -63,7 +63,7 @@ generate \
 	"${s}"
 
 # (Re)generate */Kbuild and */Config.in
-{ cd -- "$srctree" && find . -type d; } | while read -r d; do
+{ cd -- "$srctree" && find -name .\?\* -prune -or -type d -print; } | while read -r d; do
 	d="${d#./}"
 
 	src="$srctree/$d/Kbuild.src"
