--- a/Makefile	2010-11-30 09:15:30.716251554 -0800
+++ b/Makefile	2010-11-30 09:19:02.366365070 -0800
@@ -711,20 +711,10 @@ debug_kallsyms: .tmp_map$(last_kallsyms)
 endif # ifdef CONFIG_KALLSYMS
 
 # busybox image - including updated kernel symbols
-busybox_unstripped: $(busybox-all) FORCE
+busybox: $(busybox-all) FORCE
 	$(call if_changed_rule,busybox__)
 	$(Q)rm -f .old_version
 
-busybox: busybox_unstripped
-ifeq ($(SKIP_STRIP),y)
-	$(Q)cp $< $@
-else
-	$(Q)$(STRIP) -s --remove-section=.note --remove-section=.comment \
-		busybox_unstripped -o $@
-# strip is confused by PIE executable and does not set exec bits
-	$(Q)chmod a+x $@
-endif
-
 # The actual objects are generated when descending,
 # make sure no implicit rule kicks in
 $(sort $(busybox-all)): $(busybox-dirs) ;
