--- busybox.orig/applets/applets.c
+++ busybox/applets/applets.c
@@ -10,17 +10,6 @@
 #include <assert.h>
 #include "busybox.h"
 
-/* Apparently uclibc defines __GLIBC__ (compat trick?). Oh well. */
-#if ENABLE_STATIC && defined(__GLIBC__) && !defined(__UCLIBC__)
-#warning Static linking against glibc produces buggy executables
-#warning (glibc does not cope well with ld --gc-sections).
-#warning See sources.redhat.com/bugzilla/show_bug.cgi?id=3400
-#warning Note that glibc is unsuitable for static linking anyway.
-#warning If you still want to do it, remove -Wl,--gc-sections
-#warning from scripts/trylink and remove this warning.
-#error Aborting compilation.
-#endif
-
 #if ENABLE_BUILD_LIBBUSYBOX
 int main(int argc, char **argv)
 {
--- busybox.orig/scripts/trylink
+++ busybox/scripts/trylink
@@ -57,13 +57,18 @@
 EXE="$1"
 CC="$2"
 LDFLAGS="$3"
-O_FILES="$4"
-A_FILES="$5"
-LDLIBS="$6"
+LDFLAGS_STATIC="$4"
+O_FILES="$5"
+A_FILES="$6"
+LDLIBS="$7"
 
 # The -Wl,--sort-section option is not supported by older versions of ld
 SORT_SECTION=`check_cc "-Wl,--sort-section -Wl,alignment" ""`
 
+[ -z "$LDFLAGS_STATIC" ] && GC_SECTIONS="-Wl,--gc-sections"
+
+LDFLAGS="$LDFLAGS $LDFLAGS_STATIC"
+
 # Sanitize lib list (dups, extra spaces etc)
 LDLIBS=`echo "$LDLIBS" | xargs -n1 | sort | uniq | xargs`
 
@@ -76,7 +81,7 @@
 	-o $EXE \
 	-Wl,--sort-common \
 	$SORT_SECTION \
-	-Wl,--gc-sections \
+	$GC_SECTIONS \
 	-Wl,--start-group $O_FILES $A_FILES -Wl,--end-group \
 	$l_list \
 || {
@@ -100,7 +105,7 @@
 		-o $EXE \
 		-Wl,--sort-common \
 		$SORT_SECTION \
-		-Wl,--gc-sections \
+		$GC_SECTIONS \
 		-Wl,--start-group $O_FILES $A_FILES -Wl,--end-group \
 		$l_list
 	if test $? = 0; then
@@ -129,7 +134,7 @@
 	    -o $EXE \
 	    -Wl,--sort-common \
 	    $SORT_SECTION \
-	    -Wl,--gc-sections \
+	    $GC_SECTIONS \
 	    -Wl,--start-group $O_FILES $A_FILES -Wl,--end-group \
 	    $l_list \
 	    -Wl,--warn-common \
@@ -151,7 +156,7 @@
 	    -o $EXE \
 	    -Wl,--sort-common \
 	    $SORT_SECTION \
-	    -Wl,--gc-sections \
+	    $GC_SECTIONS \
 	    -Wl,-T -Wl,busybox_ldscript \
 	    -Wl,--start-group $O_FILES $A_FILES -Wl,--end-group \
 	    $l_list \
@@ -207,7 +212,7 @@
 	    -o $EXE \
 	    -Wl,--sort-common \
 	    $SORT_SECTION \
-	    -Wl,--gc-sections \
+	    $GC_SECTIONS \
 	    -Wl,--start-group $O_FILES -Wl,--end-group \
 	    -L"$sharedlib_dir" -lbusybox \
 	    -Wl,--warn-common \
@@ -246,7 +251,7 @@
 		-o $EXE \
 		-Wl,--sort-common \
 		$SORT_SECTION \
-		-Wl,--gc-sections \
+		$GC_SECTIONS \
 		-L"$sharedlib_dir" -lbusybox \
 		-Wl,--warn-common \
 	|| {
--- busybox.orig/Makefile
+++ busybox/Makefile
@@ -572,6 +572,7 @@
       "$@" \
       "$(CC)" \
       "$(LDFLAGS) $(EXTRA_LDFLAGS)" \
+      "$(LDFLAGS_STATIC)" \
       "$(core-y)" \
       "$(libs-y)" \
       "$(LDLIBS)"
--- busybox.orig/Makefile.flags
+++ busybox/Makefile.flags
@@ -66,7 +66,7 @@
 endif
 
 ifeq ($(CONFIG_STATIC),y)
-LDFLAGS += -static
+LDFLAGS_STATIC += -static
 endif
 
 LDLIBS += m crypt
